### PR TITLE
[test] `tcp-ping-pong` Does Not Wait for Full Receive Buffer

### PR DESCRIPTION
**Pull request by ppenna**
## Description

This PR fixes https://github.com/demikernel/demikernel/issues/390.

## Summary of Changes

- Fixed `tcp-ping-pong` to buffer received response on client-side
- Changed `tcp-ping-pong` to send buffers with different payload at each round.

### CODE REVIEW

File: examples/rust/tcp-ping-pong.rs [hunk 1]
```diff
@@ -68,6 +68,94 @@ fn mksga(libos: &mut LibOS, size: usize, value: u8) -> demi_sgarray_t {
...
+            Ok(qt) => qt,
+            Err(e) => panic!("pop failed: {:?}", e.cause),
+        };
+        let sga: demi_sgarray_t = match libos.wait(qt, None) {
+            Ok(qr) if qr.qr_opcode == demi_opcode_t::DEMI_OPC_POP => unsafe { qr.qr_value.sga },
+            Err(e) => panic!("operation failed: {:?}", e.cause),
+            _ => panic!("unexpected result"),
+        };
+
+        // Sanity check received data.
```
Comment (BrianZill) [inline examples/rust/tcp-ping-pong.rs:143]: Is this the right comment?  This block appears to be copying the data out to the `recvbuf`, not sanity checking it.

**Review by BrianZill — APPROVED** [review]
This change is okay (assuming this test isn't about performance).

But it brings into question what the main purpose of this test is.  If it is for correctness, then it's a reasonable test of the basics, but it doesn't check for data integrity inside each "round".  If it is for performance, then this change is going to hurt, as it introduces an extra data copy - if we care about performance here, we should hang on to the received scatter-gather arrays until we get all the data for one "round" and then sanity check the array data correctly (or perform partial checks on each receive as they arrive).  Ideally, on the server side we'd then reuse that same scatter-gather array to return that data to the client, rather than recreate it on the server side.

Even more ideally, our API would be flexible enough to allow us to modify multiple received scatter-gather arrays that compose a "round" (potentially including overlap with adjacent "rounds") to make the returning push operation take a single scatter-gather array that just describes the current "round" (without requiring any copying of the data).

File: examples/rust/tcp-ping-pong.rs (hunk 1 above)
Comment (ppenna) [reply examples/rust/tcp-ping-pong.rs]: Oh, this is indeed not the right comment. Thanks for noting, I have fixed it.

**Comment by ppenna** [comment]
> This change is okay (assuming this test isn't about performance).
> 
> But it brings into question what the main purpose of this test is. If it is for correctness, then it's a reasonable test of the basics, but it doesn't check for data integrity inside each "round". If it is for performance, then this change is going to hurt, as it introduces an extra data copy - if we care about performance here, we should hang on to the received scatter-gather arrays until we get all the data for one "round" and then sanity check the array data correctly (or perform partial checks on each receive as they arrive). Ideally, on the server side we'd then reuse that same scatter-gather array to return that data to the client, rather than recreate it on the server side.
> 
> Even more ideally, our API would be flexible enough to allow us to modify multiple received scatter-gather arrays that compose a "round" (potentially including overlap with adjacent "rounds") to make the returning push operation take a single scatter-gather array that just describes the current "round" (without requiring any copying of the data).

@BrianZill this is about fixing the correctness of the test only. Concerning the sanity check of the data in each round, we are doing that: at each round, we sanity check if the received buffer matches the `fill_char`. Are you suggesting us to have something different than that?

**Pull request merged**